### DISCUSSION TITLE
[OpenGL][Windows] How to use WGL for 3D rendering but keep Angle for Avalonia rendering?

### DISCUSSION THREAD

**Discussion by Dragorn421** [AvaloniaUI/Avalonia#16536] [Q&A]
**Under Windows, Avalonia uses Angle** for rendering controls.
But this also means that when using OpenGL **in a user control derived from `OpenGLControlBase`, the user also ends up using Angle**, which is not satisfactory for my purpose.

A "workaround" is to **tell Avalonia to use WGL**, the "native Windows OpenGL", like so:
```cs
AppBuilder.Configure<App>()
            .With(new Win32PlatformOptions()
            {
                RenderingMode = new List<Win32RenderingMode>()
                {
                    Win32RenderingMode.Wgl,
                    Win32RenderingMode.Software,
                }
            })
// ...
```
(note this does not prevent running the application on other platforms than Windows like Linux)
**However this has been dubbed as "experimental" and "not recommended"** due to Angle being "more stable" to use than Wgl.
(references: https://github.com/AvaloniaUI/Avalonia/discussions/6396 https://github.com/AvaloniaUI/Avalonia/discussions/9393 )

----------

The references I found are somewhat old, a few mention **OpenTK integration as a solution** but that plan was abandoned.
I'm not sure how OpenTK integration would have solved this anyway, wouldn't it have used the same OpenGL version/driver under the hood?
If it did not then I guess there is a possibility to keep using Angle for Avalonia rendering but use WGL for my 3D rendering, but I didn't find resources on this.

https://github.com/AvaloniaUI/Avalonia/discussions/6842#discussioncomment-1757583 mentions manual context creation with OpenGL version arguments which sounds interesting, but the classes/methods it uses are internal to Avalonia so I'm not sure what to make of it. I'm also not sure if that would still return Angle or not

-------

Hence the title-question:

**How to use WGL for 3D rendering but keep Angle for Avalonia rendering?**

Thanks :)

**Comment by maxkatz6** [comment]
> How to use WGL for 3D rendering but keep Angle for Avalonia rendering?

It's not possible. Unless you render 3D opengl into a bitmap and render this way.

> The references I found are somewhat old, a few mention OpenTK integration as a solution but that plan was abandoned.

OpenTK is a C# bindings for OpenGL. You can use them with either WGL or Angle. No special support from Avalonia is required.
There were plans to ship build-in integration, but it wasn't necessary in the end.

> https://github.com/AvaloniaUI/Avalonia/discussions/6842#discussioncomment-1757583 mentions manual context creation with OpenGL version arguments which sounds interesting,

If that's all you need, you can specify these in `.With(new AngleOptions { .. here  })` in BuildAvaloniaApp().

> but the classes/methods it uses are internal to Avalonia so I'm not sure what to make of it. 

That's a message from 2021. Pretty old :)
I think these advices evolved into GPU interop in the end, but not sure if you really need it: https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop

**Reply by Dragorn421** [reply]
Thanks for your answer

> > How to use WGL for 3D rendering but keep Angle for Avalonia rendering?
>
> It's not possible. Unless you render 3D opengl into a bitmap and render this way.

I would be fine with that, do you know of any example?
There are examples in the repo for interop with Vulkan and D3D, but not "external" OpenGL

> > https://github.com/AvaloniaUI/Avalonia/discussions/6842#discussioncomment-1757583 mentions manual context creation with OpenGL version arguments which sounds interesting,
>
> If that's all you need, you can specify these in .With(new AngleOptions { .. here  }) in BuildAvaloniaApp().

I wish, but passing an OpenGL (not ES) version to angle options wouldn't work as Angle doesn't support those versions in the first place
Avalonia even checks early for it like here https://github.com/AvaloniaUI/Avalonia/blob/3deddbe3050f67d2819d1710b2f1062b7b15868e/src/Windows/Avalonia.Win32/OpenGl/Angle/D3D11AngleWin32PlatformGraphics.cs#L63 or here https://github.com/AvaloniaUI/Avalonia/blob/3deddbe3050f67d2819d1710b2f1062b7b15868e/src/Avalonia.OpenGL/Egl/EglDisplayUtils.cs#L45

**Comment by kekekeks** [comment]
You can manually set up a WGL context, draw into ID3D11Texture2D via [WGL_NV_DX_interop2](https://registry.khronos.org/OpenGL/extensions/NV/WGL_NV_DX_interop2.txt) and then use our graphics interop APIs to import said ID3D11Texture2D into Avalonia.

It's something that we have plans for in Avalonia itself, but it's not something we currently prioritize.
